petri/petri_artifacts_core/src/lib.rs

```diff
@@ -37,14 +40,151 @@ pub trait ArtifactId: 'static {
 /// A type-safe handle to a particular Artifact, as declared using the
 /// [`declare_artifacts!`](crate::declare_artifacts) macro.
 #[derive(Copy, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
-pub struct ArtifactHandle<A: ArtifactId>(core::marker::PhantomData<A>);
+pub struct ArtifactHandle<A: ArtifactId>(PhantomData<A>);
 
 impl<A: ArtifactId + std::fmt::Debug> std::fmt::Debug for ArtifactHandle<A> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         std::fmt::Debug::fmt(&self.erase(), f)
     }
 }
 
+/// A resolved artifact path for artifact `A`.
+pub struct ResolvedArtifact<A = ()>(Option<PathBuf>, PhantomData<A>);
+
+impl<A> Clone for ResolvedArtifact<A> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone(), self.1)
+    }
+}
+
+impl<A> std::fmt::Debug for ResolvedArtifact<A> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("ResolvedArtifact").field(&self.0).finish()
+    }
+}
+
+impl<A> ResolvedArtifact<A> {
+    /// Erases the type `A`.
+    pub fn erase(self) -> ResolvedArtifact {
```

Comment:
how does this `erase` differ from `ErasedArtifactHandle`?

Reply:
This is a resolved artifact, not an artifact handle.